contexts/ProposalMarketsContext.tsx

```diff
@@ -0,0 +1,602 @@
+            if (!bids) {
+                setBids(orderBookObject?.passBidsArray);
+            }
+            if (!asks) {
+                setAsks(orderBookObject?.passAsksArray);
+            }
+            if (!spreadString) {
+                setSpreadString(orderBookObject?.passSpreadString);
+            }
+        } else {
+            if (!bids) {
+                setBids(orderBookObject?.failBidsArray);
+            }
+            if (!asks) {
+                setAsks(orderBookObject?.failAsksArray);
+            }
+            if (!spreadString) {
+                setSpreadString(orderBookObject?.failSpreadString);
+            }
+        }
+    });
+
+    const listenOrderBook = async () => {
+        if (!proposal) return;
+
+        let markets = [];
+
+        if (!isPassMarket) {
+            markets = [proposal?.account.openbookFailMarket];
+        } else {
+            markets = [proposal?.account.openbookPassMarket];
+        }
+
+        // Setup for pass and fail markets
+        markets.forEach(async (market: PublicKey) => {
+            if (!wsConnected) {
+                // Fetch via RPC for the openbook market
+                const _market = await openBookProgram.account.market.fetch(
+                    market
+                );
+                const sides = [
+                    {
+                        pubKey: _market.asks,
+                        side: 'asks',
+                    },
+                    {
+                        pubKey: _market.bids,
+                        side: 'bids',
+                    },
+                ];
+                // Setup Websocket subscription for the two sides
+                try {
+                    const subscriptionId = sides.map((side) => provider.connection.onAccountChange(
+                        side.pubKey,
+                        (updatedAccountInfo, ctx) => {
+                            consumeOrderBookSide(side.side, updatedAccountInfo, ctx);
+                        },
+                        'processed'
+                    )
+                    );
+                    return subscriptionId;
+                } catch (err) {
+                    setWsConnected(false);
+                }
+            }
+            // For map handling
+            return null;
+        });
+    };
+
+    const cancelOrder = useCallback(async (order: OpenOrdersAccountWithKey) => {
+        if (!proposal || !markets) return;
+        const marketAccount = isPassMarket
+            ? { publicKey: proposal.account.openbookPassMarket, account: markets.pass }
+            : { publicKey: proposal.account.openbookFailMarket, account: markets.fail };
+        const txs = await cancelOrderTransactions(
+            new BN(order.account.accountNum),
+            marketAccount,
+        );
+
+        if (!wallet.publicKey || !txs) return;
+
+        try {
+            //settle it right away
+            const settleTxs = await settleFundsTransactions(
```

Comment:
Is there a way to pack these into a single transaction for cancel and settle? I only ask as I'd like to get away from multi sign of transactions, or maybe I'm thinking about it the wrong way.

Reply:
I unfortunately don't see a way to do it with openbooks API but I have asked them in telegram chat if they might have a way to do it in 1 transaction, but right now not seeing it :( I was hoping I could maybe chain together the methods but that isn't happy.